### ISSUE TITLE
Issue with classification from sd-file

### ISSUE THREAD

**Issue by kienerj** [pfnet-research/chainer-chemistry#258]
Environment:
Win Server 2012 R2 64-bit
Python 3.6 (Anaconda)
CPU 

I'm trying to do a binary classification from an sd-file. The sd-file has only 1 data property, the class with values 0 or 1.

I'm following [this example](https://github.com/pfnet-research/chainer-chemistry/blob/master/examples/own_dataset/train.py) (own-dataset) adjusted for classification.

When calling `trainer.run()` I get below error:

> C:\Program Files\Anaconda3\envs\chainer\lib\site-packages\chainer\utils\type_check.py in expect(self)
>     480             raise InvalidType(
>     481                 '{0} {1} {2}'.format(self.lhs, self.exp, self.rhs),
> --> 482                 '{0} {1} {2}'.format(left, self.inv, right))
>     483 
>     484 
> 
> InvalidType: 
> Invalid operation is performed in: SigmoidCrossEntropy (Forward)
> 
> Expect: in_types[1].dtype.kind == i
> Actual: O != i

Looks like an issue with the class labels?

The code:

```
preprocessor = preprocess_method_dict['ggnn']()
parser = SDFFileParser(preprocessor, labels="Class")
dataset = parser.parse('path\to\mymolecules.sdf')["dataset"]

train_data_ratio = 0.8
seed = 777
train_data_size = int(len(dataset) * train_data_ratio)
train, val = split_dataset_random(dataset, train_data_size, seed)

n_unit = 16
conv_layers = 4
class_num = 1 # only 1 target
batch_size = 32
epochs = 20

model = GraphConvPredictor(GGNN(out_dim=n_unit, hidden_dim=n_unit,
                                        n_layers=conv_layers),
                                   MLP(out_dim=class_num, hidden_dim=n_unit))

train_iter = iterators.SerialIterator(train, batch_size)
val_iter = iterators.SerialIterator(
        val, batch_size, repeat=False, shuffle=False)

device = -1 #CPU

classifier = Classifier(model, lossfun=F.sigmoid_cross_entropy, metrics_fun=F.f1_score, device=device)

optimizer = optimizers.Adam()
optimizer.setup(classifier)

updater = training.StandardUpdater(train_iter, optimizer, device=device, converter=concat_mols)

trainer = training.Trainer(updater, (epochs, 'epoch'), out='Prediction')
trainer.extend(E.Evaluator(val_iter, classifier, device=device,
                           converter=concat_mols))
trainer.extend(E.snapshot(), trigger=(epochs, 'epoch'))
trainer.extend(E.LogReport())
# Note that original scale absolute errors are reported in
# (validation/)main/abs_error
trainer.extend(E.PrintReport(['epoch', 'main/loss', 'main/abs_error',
                              'validation/main/loss',
                              'validation/main/abs_error',
                              'elapsed_time']))
trainer.extend(E.ProgressBar())
trainer.run()

```

What am I doing wrong? Is it an issue with the class lables? I'm not seeing were I should explicitly pass them in.

**Comment by kienerj** [comment]
I changed the parser to:

`parser = SDFFileParser(preprocessor, labels=["Class"], postprocess_label=postprocess_label)`

with the postprocess method being a copy&paste from your link. I know get this different error:

`InvalidType: 
Invalid operation is performed in: ClassificationSummary (Forward)

Expect: in_types[0].shape[2:3] == in_types[1].shape[1:]
Actual: () != (1,)`

**Comment by corochann** [comment]
Does the same error happen if you use the sdf file of tox21 dataset supported in chainer chemistry?
(you can get file path of tox21 by [`get_tox21_filepath`](https://github.com/pfnet-research/chainer-chemistry/blob/master/chainer_chemistry/datasets/tox21.py#L136))

I think your code should work only changing one line `data.load_dataset`, into your sdf file loading from original tox21 training example.
https://github.com/pfnet-research/chainer-chemistry/blob/master/examples/tox21/train_tox21.py#L93

**Comment by kienerj** [comment]
I changed the dataset to:

```
from chainer_chemistry.datasets.tox21 import get_tox21_filepath
labels = ['NR-AR']

path = get_tox21_filepath('train')
preprocessor = preprocess_method_dict['ggnn']()
parser = SDFFileParser(preprocessor, labels=labels, postprocess_label=postprocess_label)
dataset = parser.parse(path)["dataset"]
```
Still get same error:

> InvalidType: 
> Invalid operation is performed in: ClassificationSummary (Forward)
> 
> Expect: in_types[0].shape[2:3] == in_types[1].shape[1:]
> Actual: () != (1,)
>

**Comment by corochann** [comment]
In that case the problem is not in sdf file itself, but your train code some how differs from `train_tox21.py` example code
https://github.com/pfnet-research/chainer-chemistry/blob/master/examples/tox21/train_tox21.py#L93

Can you run this tox21 example code above? (If it works, you can modify a little bit from this code to use your own dataset.)

**Comment by kienerj** [comment]
Ok, the issue was with the metric:

`classifier = Classifier(model, lossfun=F.sigmoid_cross_entropy, metrics_fun=F.binary_accuracy, device=device)`

I had f1_score but only binary_accuracy can be used, else the error happens.

**Comment by corochann** [comment]
Good, so you could run your training script now!? Can we close the issue?
